mil_tasks/mil_tasks_gui/dashboard.py

```diff
@@ -0,0 +1,346 @@
+        super(CenteredCheckBox, self).__init__()
+        checkbox = QtWidgets.QCheckBox()
+        layout = QtWidgets.QHBoxLayout(self)
+        layout.addWidget(checkbox)
+        layout.setAlignment(QtCore.Qt.AlignCenter)
+        layout.setContentsMargins(0, 0, 0, 0)
+        self.setLayout(layout)
+
+    def checkState(self):
+        return self.checkbox().checkState()
+
+    def setCheckState(self, state):
+        return self.checkbox().setCheckState(state)
+
+    def checked(self):
+        ''' return boolean of if box is checked '''
+        return self.checkState() == QtCore.Qt.Checked
+
+    def setChecked(self, checked):
+        ''' set checkbox to boolean checked or not '''
+        state = QtCore.Qt.Checked if checked else QtCore.Qt.Unchecked
+        return self.setCheckState(state)
+
+    def checkbox(self):
+        ''' returns internal checkbox widget for custom usage '''
+        return self.layout().itemAt(0).widget()
+
+
+class Dashboard(Plugin):
+    '''
+    RQT plugin to interface with a mil_tasks server, allowing the user to graphicly monitor
+    the current task and result, cancel the current task, run a task with optional parameters,
+    and create a linear chain of tasks.
+    '''
+    def __init__(self, context):
+        super(Dashboard, self).__init__(context)
+
+        # Create the widget and name it
+        self._widget = QtWidgets.QWidget()
+        self._widget.setObjectName("Dashboard")
+        self.setObjectName("Dashboard")
+
+        # Extend the widget with all attributes and children in the UI file
+        ui_file = os.path.join(rospkg.RosPack().get_path("mil_tasks"), "resource", "dashboard.ui")
+        loadUi(ui_file, self._widget)
+
+        self.task_runner_client = ObserveActionClient('/task', DoTaskAction)
+        self.task_runner_client.register_observer_transition_cb(self.transition_cb)
+        self.task_runner_client.register_observer_feedback_cb(self.feedback_cb)
+
+        self.current_task = None
+        self.current_task_status = ''
+        self.current_task_task = ''
+        self.current_task_result = ''
+
+        self.connect_ui()
+        self.reload_available_tasks(None)
+
+        # Deals with problem when they're multiple instances of Dashboard plugin
+        if context.serial_number() > 1:
+            self._widget.setWindowTitle(self._widget.windowTitle() + (" (%d)" % context.serial_number()))
+
+        # Add widget to the user interface
+        context.add_widget(self._widget)
+
+    def ui_log(self, string):
+        '''
+        Adds a new line to the log pane with the current time and specified string.
+
+        ex: ui_log('Hello world')
+        12:25       Hello world
+        '''
+        date_time = datetime.datetime.fromtimestamp(rospy.Time.now().to_time())
+        time_str = '{}:{}:{}'.format(date_time.hour, date_time.minute, date_time.second).ljust(12, ' ')
+        formatted = time_str + string
+        self.feedback_list.insertItem(0, formatted)
```

Comment:
Minor, but imo its more natural that the most recent log is at the bottom

Reply:
Agreed, but with Qt it is a little complicated to implement auto scrolling to the bottom. So if we insert to the bottom, the user will have to constantly scroll down to see the latest message. If you want a stab at making it work go ahead

Reply:
I gotchu:
```
From e59aabd5e45b3b16be37bfd95caada572c2b3c4a Mon Sep 17 00:00:00 2001
From: RustyBamboo <dvolya@gmail.com>
Date: Sat, 4 Nov 2017 21:20:34 -0400
Subject: [PATCH] TASKS: display most recent log at the bottom of gui

---
 mil_tasks/mil_tasks_gui/dashboard.py | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/mil_tasks/mil_tasks_gui/dashboard.py b/mil_tasks/mil_tasks_gui/dashboard.py
index 8d1a3f2..2cd2d9a 100755
--- a/mil_tasks/mil_tasks_gui/dashboard.py
+++ b/mil_tasks/mil_tasks_gui/dashboard.py
@@ -161,7 +161,14 @@ class Dashboard(Plugin):
         date_time = datetime.datetime.fromtimestamp(rospy.Time.now().to_time())
         time_str = '{}:{}:{}'.format(date_time.hour, date_time.minute, date_time.second).ljust(12, ' ')
         formatted = time_str + string
-        self.feedback_list.insertItem(0, formatted)
+
+        self.feedback_list.addItem(formatted)
+
+        # Due to Qt handling things using events, we need to keep updating the scrolling
+        self.feedback_list.model().rowsInserted.connect(self.feedback_list_signal_cb)
+
+    def feedback_list_signal_cb(self):
+        self.feedback_list.scrollToBottom()
 
     def feedback_cb(self, goal, handler, feedback):
         '''
@@ -339,6 +346,7 @@ class Dashboard(Plugin):
         self.current_task_label = self._widget.findChild(QtWidgets.QLabel, 'current_task_label')
         self.current_task_status_label = self._widget.findChild(QtWidgets.QLabel, 'current_status_label')
         self.feedback_list = self._widget.findChild(QtWidgets.QListWidget, 'log_list')
+        self.feedback_list.setAlternatingRowColors(True) # make it eaiser to read rows
         self.result_label = self._widget.findChild(QtWidgets.QLabel, 'result_label')
         self.clear_log_button = self._widget.findChild(QtWidgets.QPushButton, 'clear_log_button')
         self.clear_log_button.clicked.connect(self.clear_log)
-- 
2.7.4
```

Reply:
This would be annoying if you're trying to look at a particular part of the log while more logs are still coming in. It would constantly be scrolling to the bottom while you trying to look at somewhere above it

Reply:
Got it!